### PR TITLE
Multiple refactors

### PR DESCRIPTION
**Pull request by a1phyr**
Multiples refactors to improve `read.rs`. 

- Split a `Decompressor` type (that we could even possibly make public)
- Remove dynamic state of `ZipFile` and its `crypto_reader` field
- Simplify `make_crypto_reader` parameters

Each commit can be reviewed separately.

### CODE REVIEW

File: src/compression.rs [hunk 1]
```diff
@@ -189,6 +189,94 @@ pub const SUPPORTED_COMPRESSION_METHODS: &[CompressionMethod] = &[
...
+            Decrompressor::Zstd(r) => r.read(buf),
+            #[cfg(feature = "lzma")]
+            Decrompressor::Lzma(r) => r.read(buf),
+            #[cfg(feature = "xz")]
+            Decrompressor::Xz(r) => r.read(buf),
+        }
+    }
+}
+
+impl<R: io::BufRead> Decrompressor<R> {
```
Comment (Pr0methean) [inline src/compression.rs:228]: Typo:
```suggestion
impl<R: io::BufRead> Decompressor<R> {
```

File: src/read.rs [hunk 2]
```diff
@@ -1456,21 +1323,8 @@ pub(crate) fn parse_single_extra_field<R: Read>(
...
-        Ok(&mut self.reader)
-    }
-
-    pub(crate) fn get_raw_reader(&mut self) -> &mut dyn Read {
-        if let ZipFileReader::NoReader = self.reader {
-            let crypto_reader = self.crypto_reader.take().expect("Invalid reader state");
-            self.reader = ZipFileReader::Raw(crypto_reader.into_inner())
-        }
-        &mut self.reader
+    pub(crate) fn get_raw_reader(&mut self) -> io::Take<&'a mut dyn Read> {
```
Comment (Pr0methean) [inline src/read.rs:1326]: To make the side effect on self more obvious:
```suggestion
    pub(crate) fn take_raw_reader(&mut self) -> io::Take<&'a mut dyn Read> {
```

File: src/read.rs [hunk 3]
```diff
@@ -184,78 +168,32 @@ impl<'a> CryptoReader<'a> {
...
-    Lzma(Crc32Reader<Box<LzmaDecoder<CryptoReader<'a>>>>),
-    #[cfg(feature = "xz")]
-    Xz(Crc32Reader<XzDecoder<CryptoReader<'a>>>),
+    Compressed(Box<Crc32Reader<Decrompressor<io::BufReader<CryptoReader<'a>>>>>),
 }
 
 impl<'a> Read for ZipFileReader<'a> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         match self {
             ZipFileReader::NoReader => panic!("ZipFileReader was in an invalid state"),
```
Comment (Pr0methean) [inline src/read.rs:177]: This crate is meant to be panic-free even when misused.
```suggestion
            ZipFileReader::NoReader => io::Error::other("ZipFileReader is already closed or not opened"),
```

**Review by Pr0methean — CHANGES_REQUESTED** [review]
Looks good except for a few nitpicks.

File: src/read.rs (hunk 3 above)
Comment (a1phyr) [reply src/read.rs]: The panic was already there but I agree that an error would be better

File: src/read.rs [hunk 4]
```diff
@@ -199,141 +183,69 @@ impl<'a> CryptoReader<'a> {
     }
 }
 
+#[cold]
+fn invalid_state() -> io::Error {
```
Comment (Pr0methean) [inline src/read.rs:187]: Make this return `io::Result<T>` instead, since it's always wrapped.

**Review by Pr0methean — CHANGES_REQUESTED** [review]
Thanks again for this! Just one more nitpick.

File: src/read.rs [hunk 5]
```diff
@@ -199,141 +183,69 @@ impl<'a> CryptoReader<'a> {
...
+    io::Error::new(
+        io::ErrorKind::Other,
+        "ZipFileReader was in an invalid state",
+    )
+}
```
Comment (Pr0methean) [inline src/read.rs:192]:
```suggestion
fn invalid_state<T>() -> io::Result<T> {
    Err(io::Error::new(
        io::ErrorKind::Other,
        "ZipFileReader was in an invalid state",
    ))
}
```

File: src/read.rs [hunk 6]
```diff
@@ -199,141 +183,69 @@ impl<'a> CryptoReader<'a> {
...
-    #[cfg(feature = "xz")]
-    Xz(Crc32Reader<XzDecoder<CryptoReader<'a>>>),
+    Compressed(Box<Crc32Reader<Decompressor<io::BufReader<CryptoReader<'a>>>>>),
 }
 
 impl<'a> Read for ZipFileReader<'a> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         match self {
-            ZipFileReader::NoReader => panic!("ZipFileReader was in an invalid state"),
+            ZipFileReader::NoReader => Err(invalid_state()),
```
Comment (Pr0methean) [inline src/read.rs:203]:
```suggestion
            ZipFileReader::NoReader => invalid_state(),
```

File: src/read.rs [hunk 7]
```diff
@@ -199,141 +183,69 @@ impl<'a> CryptoReader<'a> {
...
-            #[cfg(feature = "xz")]
-            ZipFileReader::Xz(r) => r.read(buf),
+            ZipFileReader::Compressed(r) => r.read(buf),
         }
     }
 
     fn read_exact(&mut self, buf: &mut [u8]) -> io::Result<()> {
         match self {
-            ZipFileReader::NoReader => panic!("ZipFileReader was in an invalid state"),
+            ZipFileReader::NoReader => Err(invalid_state()),
```
Comment (Pr0methean) [inline src/read.rs:211]:
```suggestion
            ZipFileReader::NoReader => invalid_state(),
```

File: src/read.rs [hunk 8]
```diff
@@ -199,141 +183,69 @@ impl<'a> CryptoReader<'a> {
...
-            #[cfg(feature = "xz")]
-            ZipFileReader::Xz(r) => r.read_exact(buf),
+            ZipFileReader::Compressed(r) => r.read_exact(buf),
         }
     }
 
     fn read_to_end(&mut self, buf: &mut Vec<u8>) -> io::Result<usize> {
         match self {
-            ZipFileReader::NoReader => panic!("ZipFileReader was in an invalid state"),
+            ZipFileReader::NoReader => Err(invalid_state()),
```
Comment (Pr0methean) [inline src/read.rs:219]:
```suggestion
            ZipFileReader::NoReader => invalid_state(),
```

File: src/read.rs [hunk 9]
```diff
@@ -199,141 +183,69 @@ impl<'a> CryptoReader<'a> {
...
-            #[cfg(feature = "xz")]
-            ZipFileReader::Xz(r) => r.read_to_end(buf),
+            ZipFileReader::Compressed(r) => r.read_to_end(buf),
         }
     }
 
     fn read_to_string(&mut self, buf: &mut String) -> io::Result<usize> {
         match self {
-            ZipFileReader::NoReader => panic!("ZipFileReader was in an invalid state"),
+            ZipFileReader::NoReader => Err(invalid_state()),
```
Comment (Pr0methean) [inline src/read.rs:227]:
```suggestion
            ZipFileReader::NoReader => invalid_state(),
```

File: src/read.rs [hunk 10]
```diff
@@ -199,141 +183,69 @@ impl<'a> CryptoReader<'a> {
...
-            ZipFileReader::Lzma(r) => {
-                // Lzma reader owns its buffer rather than mutably borrowing it, so we have to drop
-                // it separately
-                if let Ok(mut remaining) = r.into_inner().finish() {
-                    let _ = copy(&mut remaining, &mut sink());
-                }
-                return;
+    fn into_inner(self) -> io::Result<io::Take<&'a mut dyn Read>> {
+        match self {
+            ZipFileReader::NoReader => Err(invalid_state()),
```
Comment (Pr0methean) [inline src/read.rs:237]:
```suggestion
            ZipFileReader::NoReader => invalid_state(),
```

**Review by Pr0methean — APPROVED** [review]

**Pull request merged**